deepspeed/checkpoint/reshape_utils.py

```diff
@@ -0,0 +1,87 @@
+import os
+import torch
+from collections import OrderedDict
+
+
+def basic_folder_validation(dir):
+    assert os.path.exists(dir), f'{dir} path does not exist'
+    assert os.path.isdir(dir), f'{dir} is not a folder'
+
+
+def get_files_with_prefix(all_files, prefix):
+    file_list = []
+    for file_path in all_files:
+        _, fname = os.path.split(file_path)
+        if fname.startswith(prefix):
+            file_list.append(file_path)
+
+    return sorted(file_list)
+
+
+def validate_files(file_list):
+    for file in file_list:
+        if not os.path.isfile(file):
+            print(f'Error: {file} is not existent')
+
+
+def get_files(dir):
+    file_list = []
+    for root, _, files in os.walk(dir):
+        for file in files:
+            file_list.append(os.path.join(root, file))
+    return file_list
+
+
+def partition_data(data_list, num_partitions):
+    num_elems = len(data_list)
+    assert num_elems % num_partitions == 0
+    partition_size = num_elems // num_partitions
+    partitions_list = [
+        data_list[i:i + partition_size] for i in range(0,
+                                                       num_elems,
+                                                       partition_size)
+    ]
+    return partitions_list
+
+
+def _key_list_to_string(key_list):
+    return '.'.join(key_list)
+
+
+def merge_state_dict(dict_a, dict_b, key_list):
+    merged_dict = type(dict_a)({})
+
+    for key, value in dict_b.items():
+        if key in dict_a.keys():
+            merged_dict[key] = merge_state(dict_a[key], dict_b[key], [str(key)])
+        else:
+            merged_dict[key] = value
+
+    return merged_dict
+
+
+def merge_state_list(list_a, list_b, key_list):
+    if len(list_a) != len(list_b):
+        print(f'{_key_list_to_string(key_list)}')
+        raise ValueError(
+            f'Cannot merge lists of different lengths, a = {len(list_a)} b = {len(list_b)}'
+        )
+
+    return [merge_state(a, b, key_list) for a, b in zip(list_a, list_b)]
+
+
+def merge_state(state_a, state_b, key_list=[]):
+    if type(state_a) != type(state_b):
+        key_list_string = _key_list_to_string(key_list)
+        print(f'key_list = {key_list_string}')
+        raise ValueError(
+            f'Cannot merge two states of types {type(state_a)} and type {type(state_b)}')
+
+    if type(state_a) in (dict, OrderedDict):
+        return merge_state_dict(state_a, state_b, key_list)
+    elif type(state_a) in (list, tuple):
+        return type(state_a)(merge_state_list(state_a, state_b, key_list))
+    elif torch.is_tensor(state_a):
+        return torch.cat([state_a, state_b], 0)
+    else:
+        return state_a
```

Comment:
Maybe it'd be nice to have an arg to pass a list of keys where always `state_a` should be taken?
Currently, this merges RNG states for me, which is undesirable

Reply:
This is correct. I had plans to do this as I gained more understanding of the various merge semantics but never got around to it. In general, it seems there are three classes of merge: (1) concat, (2) pick any (or `state_a` in this case) , and (3) average. So, my thought was to pass in some sort of mapping of keys to one of these three merge strategies. In practice, it seems (1) and (2) only apply for Megatron-DeepSpeed. Is this useful information? 

Also, please note that we somewhat abandoned this approach in favor of the universal checkpointing implemented [here](https://github.com/bigscience-workshop/Megatron-DeepSpeed/pull/239/files#diff-6c7e4f412a4bdd7ccd977eaacfcf8eb7bedfffb0c08e65776c00df9aced4f129).  You can see the merging logic for universal approach [here](https://github.com/bigscience-workshop/Megatron-DeepSpeed/blob/a247614b2fa41326ac8482a1735df867778cec85/tools/convert_checkpoint/ds_to_universal.py#L233-L239).

Reply:
Nice, that approach is a lot cleaner.
Do you think it's a big problem we're not averaging the layernorms in the above & thus [here](https://github.com/bigscience-workshop/Megatron-DeepSpeed/pull/239/files#diff-ecfdbe3a107133adc51e3f7c5d73ef006d6b06bd18bfc16ecb4708834e6fdc97)?
My assumption is that as long as we continue training afterwards it should be completely fine

Reply:
Yes, I believe we skipped the averaging because the layernorms in the Big Science model get automatically fixed when we continue training. But we should probably average to generalize to other models, so this needs a TODO.

Reply:
layernorms should be identical across TPs. Except they aren't due a bug somewhere in Meg-DS which only manifests at huge models. Which we workaround by forcefully resyncing layernorms on each forward path using this branch https://github.com/bigscience-workshop/Megatron-DeepSpeed/pull/272 so it should be safe to just pick the first one.